smelli/unbinned.py

```diff
@@ -0,0 +1,111 @@
+from flavio.io import instanceio as iio
+import numpy as np
+import yaml
+from .util import get_cachepath
+from requests import get
+import os
+from wilson import Wilson
+from sklearn.neighbors import KernelDensity
+import wcxf
+
+class UnbinnedParameterLikelihood(iio.YAMLLoadable):
+    """An `UnbinnedParameterLikelihood` provides an unbinned likelihood function in terms of
+    parameters.
+    Methods:
+    - `log_likelihood`: The likelihood as a function of the parameters
+    Instances can be imported and exported from/to YAML using the `load`
+    and `dump` methods.
+    """
+    _input_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': str,
+        'weights': str
+    }
+
+    _output_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': str,
+        'weights': str
+    }
+
+    def __init__(self,
+                 eft,
+                 basis,
+                 scale,
+                 names,
+                 samples,
+                 base_path,
+                 weights=None):
+        """Initialize the instance.
+        Parameters:
+        - `eft`: a string identifying the WET as defined by wcxf
+        - `basis`: a string identifying the WET basis as defined by wcxf
+        - `scale`: a scale (in units of GeV) at which the WET coefficient samples are valid
+        - `names`: an ordered list of names identiying the Wilson coefficients in the WET basis
+        - `samples`: a numpy array of vectors of Wilson coefficients, with each row representing a sample
+        - `weights`: a numpy array of scalar weights on a logarithmic scale for each sample
+        """
+        self.eft     = eft
+        self.basis   = basis
+        self.scale   = scale
+        self.names   = names
+        self.samples = np.load(os.path.join(base_path, samples))
+        self.weights = np.ones(len(self.samples)) if weights is None else np.load(os.path.join(base_path, weights))
+        self.kde     = KernelDensity(algorithm='auto', kernel='gaussian', rtol=1e-4)
+
+        basis = wcxf.Basis[eft, basis]
+        basis_wcs = {}
+        for sector in basis.sectors.values():
+            basis_wcs.update(sector)
+
+        self.accessors = []
+        for i, n in enumerate(names):
+            if 'real' in basis_wcs[n]:
+                self.accessors.append(lambda wcs: wcs[i])
+            else:
+                self.accessors.append(lambda wcs: np.real(wcs[i]))
+                self.accessors.append(lambda wcs: np.imag(wcs[i]))
+
+        self.samples = np.array([
+            [a(row) for a in self.accessors] for row in self.samples
+        ])
```

Comment:
Instead of using the accessors (which might not work, see my comment above), one could do something like this:

Outside of class, import chain function
```python
from itertools import chain
```

Inside `__init__` method of `UnbinnedParameterLikelihood` class, make `basis_wcs` a private attribute
```python
self.basis_wcs = basis_wcs
```

Inside `UnbinnedParameterLikelihood` class, define utility method for converting list of n complex values and m real values to list of 2n+m real values given that `self.names` contains n+m names that can be found in `self._basis_wcs`
```python
def _complex_to_real(self, values):
    return list(chain.from_iterable(
        [np.real(values[i])] if 'real' in self._basis_wcs[n]
        else
        [np.real(values[i]), np.imag(values[i])]
        for i, n in enumerate(self.names)
    ))
```
convert the samples using
```python
self.samples = np.array([
    self._complex_to_real(row) for row in self.samples
])
```
Use the `_complex_to_real` method also later inside the ` log_likelihood` method on `values`:
```python
values = self._complex_to_real([wet_wc[name] for name in self.names])
```

Reply:
Sorry Peter, I did not see this comment before just now, after fixing my approach. Reading up on ```chain``` right now.

Reply:
`chain` can be used to flatten a list:
```python
list(chain.from_iterable([[1], [2,3], [4,5], [6], [7]]))
```
returns
```
[1, 2, 3, 4, 5, 6, 7]
```
So one can put real values in single-element lists and complex values in two-element lists and then flatten the list using `chain`.